lib/vagrant-libvirt/config.rb

```diff
@@ -659,10 +659,26 @@ def finalize!
         @management_network_domain = nil if @management_network_domain == UNSET_VALUE
         @system_uri      = 'qemu:///system' if @system_uri == UNSET_VALUE
 
+        # Default to false for qemu_use_session
         @qemu_use_session = false if @qemu_use_session == UNSET_VALUE
 
-        # generate a URI if none is supplied
-        @uri = _generate_uri if @uri == UNSET_VALUE
+        # If uri is not set then:
+        #   - set based on LIBVIRT_DEFAULT_URI in env if exists
+        #   - OR generate it from scratch
+        if @uri == UNSET_VALUE
+          if ENV['LIBVIRT_DEFAULT_URI']
+            @uri = ENV['LIBVIRT_DEFAULT_URI']
+            # If we're setting the URI from the env var then let's
+            # appropriately update @qemu_use_session based on the @uri
+            if (@uri.include? "qemu") && (@uri.include? "session")
+              @qemu_use_session = true
+            else
+              @qemu_use_session = false
+            end
```

Comment:
Should only do this if `@qemu_use_session` has not been explicitly set by configuration, Probably easier to to split this if/else block to just an if and leave the original statement to set `@uri` as it was below. Then you can move the `@qemu_use_session` setting line from above this block to below just before resolving `@uri` if not explicitly set. That way can test at this point whether `@qemu_use_session == UNSET_VALUE` before trying to infer the value from the current contents.

Of course possibly it might be a good idea to move all of this into a separate method and that way we could easily add some tests just for that method to check that the different scenarios are correct covered to avoid surprises.

Reply:
ok I think I've addressed your concerns by moving the inspecting of the environment var into the `_generate_uri` function. I don't have a lot of time so figuring out the tests would be something I'd appreciate some help on if you don't mind.